File: library/src/main/java/renetik/java/collections/CSLinkedMapImpl.java
package renetik.java.collections;

import java.util.Map;

import static renetik.android.lang.CSLang.list;

public class CSLinkedMapImpl<K, V> extends java.util.LinkedHashMap<K, V> implements CSLinkedMap<K, V> {

    public CSLinkedMapImpl(Map<? extends K, ? extends V> m){
        super(m);
    }

    public CSLinkedMapImpl(){
    }


    public V getValue(int index) {
        return list(values()).get(index);
    }

    public int index(K key) {
        return list(keySet()).index(key);
    }

    public boolean hasKey(K key) {
        return super.containsKey(key);
    }

    public boolean hasValue(V value) {
        return super.containsValue(value);
    }

    public V value(K key) {
        return super.get(key);
    }

    public V removeAt(int i) {
        return remove(list(keySet()).get(i));
    }
}

File: library/renetik-base/src/main/java/renetik/android/java/collections/CSLinkedMapImpl.kt
package renetik.android.java.collections

class CSLinkedMapImpl<K, V> : java.util.LinkedHashMap<K, V>, CSLinkedMap<K, V> {

    constructor(m: Map<out K, V>) : super(m) {}

    constructor() {}

    override fun getValue(index: Int) = list(values)[index]

    override fun index(key: K) = list(keys).indexOf(key)

    override fun hasKey(key: K) = super.containsKey(key)

    override fun hasValue(value: V) = super.containsValue(value)

    override fun value(key: K) = super.get(key)

    fun removeAt(i: Int) = remove(list(keys)[i])
}
